openff/toolkit/topology/molecule.py

```diff
@@ -1380,31 +1437,42 @@ def delete_hierarchy_scheme(self, iter_name):
                 f"because no HierarchyScheme with that iterator name exists"
             )
         self._hierarchy_schemes.pop(iter_name)
-        if hasattr(self, iter_name):
-            delattr(self, iter_name)
 
-    def perceive_hierarchy(self, iter_names=None):
+    def update_hierarchy_schemes(self, iter_names=None):
         """
-        Infer a hierarchy from atom metadata according to the existing hierarchy schemes.
+        Infer a hierarchy from atom metadata according to the existing hierarchy
+        schemes.
+
+        Hierarchy schemes allow iteration over groups of atoms according to
+        their metadata. For more information, see
+        :class:`~openff.toolkit.topology.molecule.HierarchyScheme`.
 
         Parameters
         ----------
         iter_names : Iterable of str, Optional
-            Only perceive hierarchy for HierarchySchemes that expose these iterator names.
-            If not provided, all known hierarchies will be perceived, overwriting previous
-            results if applicable.
+            Only perceive hierarchy for HierarchySchemes that expose these
+            iterator names. If not provided, all known hierarchies will be
+            perceived, overwriting previous results if applicable.
         """
         if iter_names is None:
             iter_names = self._hierarchy_schemes.keys()
 
         for iter_name in iter_names:
             hierarchy_scheme = self._hierarchy_schemes[iter_name]
             hierarchy_scheme.perceive_hierarchy()
-            self._expose_hierarchy_scheme(iter_name)
 
-    def _expose_hierarchy_scheme(self, iter_name):
-        assert iter_name in self._hierarchy_schemes
-        setattr(self, iter_name, self._hierarchy_schemes[iter_name].hierarchy_elements)
+    def __getattr__(self, name: str):
+        """If a requested attribute is not found, check the hierarchy schemes"""
+        try:
+            return self.__dict__["_hierarchy_schemes"][name].hierarchy_elements
```

Comment:
(not blocking) I see that calling to `__dict__` in `__getattr__` is the recommended use, however this spooked me a little bit because I hadn't considered the difference between `__dict__` and `to_dict`. 

The way that a lot of our serialization works is that we call `to_dict`, which recursively converts the object and everything it "owns" to a serializable representation (consisting only of dicts, lists, tuples, ints, strings, and floats). Some things that are attributes of an object aren't "owned" by the object. For example, I can call `hierarchy_scheme.parent` and get a Molecule, but when I call `hierarchy_scheme.to_dict()`, there is no `parent` in the keys of the resulting dict. The deliberate separation of "attributes" from "things that get serialized in `to_dict`" is important to prevent infinite recursion (which I see is causing problems recently, though this particular situation may not be the root cause). 

It would be bad if `to_dict` were called here, since then this wouldn't return a list of `HierarchyElements`, instead it would return a list of _serialized_ `HierarchyElements` (that is, dicts). 

So, in the end, I think I've convinced myself that this is OK, and probably even correct. Thanks for iterating on this!

Reply:
Yeah I think `__dict__` and `to_dict` are two completely different things - the former is just the dict that every Python object uses to store its attributes, and the latter is for serialization.